File: src/main/java/de/ingrid/mapclient/Configuration.java
/*
 * **************************************************-
 * InGrid Web Map Client
 * ==================================================
 * Copyright (C) 2014 - 2016 wemove digital solutions GmbH
 * ==================================================
 * Licensed under the EUPL, Version 1.1 or – as soon they will be
 * approved by the European Commission - subsequent versions of the
 * EUPL (the "Licence");
 * 
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 * 
 * http://ec.europa.eu/idabc/eupl5
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 * **************************************************#
 */
/*
 * Copyright (c) 2011 wemove digital solutions. All rights reserved.
 */
package de.ingrid.mapclient;

import java.util.ArrayList;
import java.util.List;

import de.ingrid.mapclient.model.AreaCategory;
import de.ingrid.mapclient.model.Layer;
import de.ingrid.mapclient.model.MapExtend;
import de.ingrid.mapclient.model.Projection;
import de.ingrid.mapclient.model.Scale;
import de.ingrid.mapclient.model.ServiceCategory;
import de.ingrid.mapclient.model.Setting;

/**
 * Configuration holds the dynamic configuration of the map client that
 * is managed in the administration application.
 * 
 * @author ingo@wemove.com
 */
public class Configuration {

	private String wmsCapUrl;
	private String featureUrl;
	private String wmsCopyright;

	private List<Layer> layers;
	private MapExtend mapExtend;
	private List<Projection> projections;
	private List<Scale> scales;
	private List<Setting> settings;
	private String proxyUrl;
	private List<ServiceCategory> serviceCategories;
	private List<AreaCategory> areaCategories;

	/**
	 * Constructor
	 */
	public Configuration() {
		this.wmsCapUrl = "";
		this.wmsCopyright = "";
		this.layers = new ArrayList<Layer>();
		this.mapExtend = new MapExtend();
		this.projections = new ArrayList<Projection>();
		this.scales = new ArrayList<Scale>();
		this.setSettings(new ArrayList<Setting>());
		this.proxyUrl = "";
	}

	public String getWmsCapUrl() {
		return this.wmsCapUrl;
	}

	public void setWmsCapUrl(String wmsCapUrl) {
		this.wmsCapUrl = wmsCapUrl;
	}

	public List<Layer> getLayers() {
		return this.layers;
	}

	public void setLayers(List<Layer> layers) {
		this.layers = layers;
	}

	public MapExtend getMapExtend() {
		return this.mapExtend;
	}

	public void setMapExtend(MapExtend mapExtend) {
		this.mapExtend = mapExtend;
	}

	public List<Projection> getProjections() {
		return this.projections;
	}

	public void setProjections(List<Projection> projections) {
		this.projections = projections;
	}

	public List<Scale> getScales() {
		return this.scales;
	}

	public void setScales(List<Scale> scales) {
		this.scales = scales;
	}

	public String getProxyUrl() {
		return this.proxyUrl;
	}

	public void setProxyUrl(String proxyUrl) {
		this.proxyUrl = proxyUrl;
	}

	public List<ServiceCategory> getServiceCategories() {
		return this.serviceCategories;
	}

	public void setServiceCategories(List<ServiceCategory> serviceCategories) {
		this.serviceCategories = serviceCategories;
	}

	public List<AreaCategory> getAreaCategories() {
		return this.areaCategories;
	}

	public void setAreaCategeories(List<AreaCategory> areaCategories) {
		this.areaCategories = areaCategories;
	}
	public String getFeatureUrl() {
		return featureUrl;
	}

	public void setFeatureUrl(String featureUrl) {
		this.featureUrl = featureUrl;
	}

	public String getWmsCopyright() {
		return wmsCopyright;
	}

	public void setWmsCopyright(String wmsCopyright) {
		this.wmsCopyright = wmsCopyright;
	}

	public List<Setting> getSettings() {
		return settings;
	}

	public void setSettings(List<Setting> settings) {
		this.settings = settings;
	}
}

File: src/main/webapp/shared/js/Configuration.js
/*
 * **************************************************-
 * InGrid Web Map Client
 * ==================================================
 * Copyright (C) 2014 wemove digital solutions GmbH
 * ==================================================
 * Licensed under the EUPL, Version 1.1 or – as soon they will be
 * approved by the European Commission - subsequent versions of the
 * EUPL (the "Licence");
 * 
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 * 
 * http://ec.europa.eu/idabc/eupl5
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 * **************************************************#
 */
/*
 * Copyright (c) 2011 wemove digital solutions. All rights reserved.
 */
Ext.namespace("de.ingrid.mapclient");

/**
 * @class Configuration holds all configuration parameters and gives access to them.
 * On application startup the de.ingrid.mapclient.Configuration.load() method must
 * be called to initialize the configuration.
 *
 * @constructor
 */
de.ingrid.mapclient.Configuration = function() {

	/**
	 * The static configuration
	 */
	this.staticConfig = {};

	/**
	 * The dynamic configuration
	 */
	this.dynamicConfig = {};

	/**
	 * Indicates if the configuration is loaded
	 */
	this.isLoaded = false;
	
	/**
	 * hide (or not) experimental features
	 */
	this.hiddenFeature = true;
};

/**
 * Load the configuration from the server
 * @param responseHandler Object with properties success and failure, which are both functions
 * 		to be called with the response text as parameter in the appropriate case (optional)
 */
de.ingrid.mapclient.Configuration.load = function(responseHandler) {
	
	de.ingrid.mapclient.Configuration.instance = new de.ingrid.mapclient.Configuration();
	// load the dynamic configuration
	Ext.Ajax.request({
		url: de.ingrid.mapclient.PERSISTENT_DYNAMIC_CONFIG_BASE_URL,
		method: 'GET',
		success: function(response, request) {
			// merge the configuration values into the instance
			var config = Ext.decode(response.responseText);
			Ext.apply(de.ingrid.mapclient.Configuration.instance.dynamicConfig, config);

			// load the static configuration
			Ext.Ajax.request({
				url: de.ingrid.mapclient.STATIC_CONFIG_BASE_URL,
				method : 'GET',
				success : function(response, request) {
					// merge the configuration values into the instance
					var config = Ext.decode(response.responseText);
					for (var i=0, count=config.length; i<count; i++) {
						var curValue = config[i];
						de.ingrid.mapclient.Configuration.instance.staticConfig[curValue['@name']] = curValue['@value'];
					}

					// set the load status
					de.ingrid.mapclient.Configuration.instance.isLoaded = true;

					// call the callback
					if (responseHandler.success instanceof Function) {
						responseHandler.success(response.responseText);
					}
				}
			});
		},
		failure: function(response, request) {
			if (responseHandler.failure instanceof Function) {
				responseHandler.failure(response.responseText);
			}
		}
	});
};

/**
 * Get a value from the dynamic configuration
 * @param key The name of the value
 * @return Object
 */
de.ingrid.mapclient.Configuration.getValue = function(key) {
	this.checkLoaded();
	return de.ingrid.mapclient.Configuration.instance.dynamicConfig[key];
};

/**
 * Set a value in the configuration and store it to the server. Each value is
 * supposed to be stored by sending a POST request to DYNAMIC_CONFIG_BASE_URL+'/'+key
 * @param key The name of the value
 * @param value The value as JSON encoded string
 * @param responseHandler Object with properties success and failure, which are both functions
 * 		to be called with the response text as parameter in the appropriate case (optional)
 */
de.ingrid.mapclient.Configuration.setValue = function(key, value, responseHandler) {
	this.checkLoaded();
	Ext.Ajax.request({
		url: de.ingrid.mapclient.DYNAMIC_CONFIG_BASE_URL+'/'+key,
		method: 'POST',
		headers: { 'Content-Type': 'text/plain' },
		success: function(response, request) {
			if (responseHandler.success instanceof Function) {
				responseHandler.success(response.responseText);
			}
			// only change local value on success
			de.ingrid.mapclient.Configuration.instance.dynamicConfig[key] = value;
		},
		failure: function(response, request) {
			if (responseHandler.failure instanceof Function) {
				responseHandler.failure(response.responseText);
			}
		},
		xmlData: value
	});
};

/**
 * Get a value from the static configuration
 * @param key The name of the value
 * @return String
 */
de.ingrid.mapclient.Configuration.getProperty = function(key) {
	this.checkLoaded();
	return de.ingrid.mapclient.Configuration.instance.staticConfig[key];
};

/**
 * Get a parameter from the application url
 * @param key The name of the parameter
 * @return String
 */
de.ingrid.mapclient.Configuration.getUrlParameter = function(key) {
	var getParams = document.URL.split("?");
	if (getParams.length > 1) {
		var params = Ext.urlDecode(getParams[getParams.length-1]);
		if (params[key]) {
			return params[key];
		}
	}
	return undefined;
};

/**
 * Check if the configuration is loaded and throw an exception if not
 */
de.ingrid.mapclient.Configuration.checkLoaded = function() {
	var instance = de.ingrid.mapclient.Configuration.instance;
	if (!instance || !instance.isLoaded) {
		throw "The configuration is not loaded.";
	}
};


/**
 * Get Settings from Admin
 */
de.ingrid.mapclient.Configuration.getSettings = function(key) {
	var settings = de.ingrid.mapclient.Configuration.getValue('settings');
	if(settings){
		for (var j=0, countJ=settings.length; j<countJ; j++) {
			var setting = settings[j];
			if(setting.key == key){
				var value = setting.value;
				// Check if boolean value
				if (value.toLowerCase()=='false'){
					value = false;
				} else if (value.toLowerCase()=='true'){
				    value =  true;
				}
				return value;
			}
		}
	}
};